### PR TITLE
backport: S3C-1640 CRR retry feature

### PR DESCRIPTION
**Pull request by jonathan-gramain**
Depends on https://github.com/scality/backbeat/pull/406

backport: S3C-1640 CRR retry feature

backport the following commits from development/8.1:

* ec16f1f4 FT: Add failure retry routes
* e74e871f FT: Add retry support using kafka and redis
* bb45ac55 rf: ZENKO-147 Refactor retry CRR API
* 43402552 bf: ZENKO-147 Check site when setting failed hash
* cdb417b3 ft: ZENKO-147 delete hash field on retry
* 45c9d419 ft: ZENKO-147 Make CRR retry configurable
* 7d06d0ea ft: S3C-1398 Use retry CRR topic
* 606c0f54 improvement: ZENKO-353 Avoid storing all objectMD
* 2da08f9e ft: ZENKO-147 Use Redis keys instead of hash
* a145a16b rf: ZENKO-143 move attachReqUids() helper
* 3abaaaed bugfix: ZENKO-1144 Use sorted sets for failures
* 7c31e958 bugfix: ZENKO-1144 fix edge cases
* b4568f6a bugfix: ZENKO-1144: Improve listing response time
* 84b95f72 bugfix: ZENKO-1024 Enhance functional tests

partial backport of the following commit from development/8.0:

* 0db307b9 ft: add site-level metrics

Note: only the changes necessary to use the "_redisClient" field in
BackbeatAPI.js have been backported from 0db307b9.

Manual test done:

* shutdown target S3C site
* put a small object on a replicated bucket
* wait 5mn until status becomes FAILED
* do a curl request on http://localhost:8900/_/crr/failed: the FAILED entry appears in the result
* restart target S3C site
* do a POST request on the same URL: the entry becomes PENDING
* the entry replicates and status becomes SUCCESS

### CODE REVIEW

**Comment by philipyoo** [comment]
@jonathan-gramain Will there be issues if we backport the failed crr producer/consumer? If not, I can look into including those changes as well

**Comment by bennettbuchanan** [comment]
@jonathan-gramain We had refactored this feature quite a bit last month due to a scaling issue related to failure metrics reporting. Do you know, will these CRR metrics functionality be needed in S3C as well? If so, there are a number of additional considerations related to the route's usage in S3C.

**Comment by jonathan-gramain** [comment]
@philipyoo @bennettbuchanan I think we should include all fixes that we have done on CRR retry and related topics (so CRR stats could be one if linked in some way to CRR retry).

@philipyoo go ahead if you want to backport failed CRR producer/consumer, I think it will be needed.

File: lib/api/BackbeatAPI.js [hunk 1]
```diff
@@ -77,7 +111,8 @@ class BackbeatAPI {
         if (route.substring(3) === 'healthcheck') {
             return true;
         }
-
+        const backbeatRoutes = getRoutesFn(redisKeys, this._validSites);
+        routes.push(...backbeatRoutes);
```
Comment (philipyoo) [inline lib/api/BackbeatAPI.js]: I think this will actually push the routes from `getRoutesFn` each request. An alternative can be moving this to the constructor

File: lib/api/BackbeatAPI.js [hunk 2]
```diff
@@ -54,14 +66,36 @@ class BackbeatAPI {
         } else {
             this._internalStart = Date.now();
         }
-
-        const redisClient = new RedisClient(this._redisConfig, this._logger);
+        // Redis instance for publishing messages to BackbeatConsumers
+        this._redisPublisher = new Redis(this._redisConfig);
```
Comment (philipyoo) [inline lib/api/BackbeatAPI.js]: we can remove this, only used for pause/resume atm

File: lib/api/BackbeatRequest.js [hunk 3]
```diff
@@ -24,6 +26,33 @@ class BackbeatRequest {
...
+            const { versionId, marker, sitename } = querystring.parse(query);
+            this._routeDetails.extension = parts[0];
+            this._routeDetails.status = parts[1];
+            this._routeDetails.bucket = parts[2];
+            this._routeDetails.key = parts.slice(3).join('/');
+            this._routeDetails.versionId = versionId;
+            this._routeDetails.marker = marker;
+            this._routeDetails.sitename = sitename;
+        } else {
+            // for now: pause/resume/status
```
Comment (philipyoo) [inline lib/api/BackbeatRequest.js]: can remove this else block

**Review by philipyoo — COMMENTED** [review]
routes lgtm, a few comments though 👍

File: lib/api/BackbeatAPI.js (hunk 1 above)
Comment (bennettbuchanan) [reply lib/api/BackbeatAPI.js]: Removed this and back-ported routes from Arsenal into routes.js instead.

**Review by philipyoo — APPROVED** [review]

**Review by jonathan-gramain — COMMENTED** [review]
E2E have passed (manual run on custom Integration branch test/backportCRRRetryZenkoE2ETests)

I cannot approve my own PR, but :shipit:

**Review by bennettbuchanan — APPROVED** [review]
Approving in the spirit of @jonathan-gramain
